src/ucp/api/ucp_def.h

```diff
@@ -771,5 +771,45 @@ typedef struct ucp_ep_params {
  */
 #define UCP_ENTITY_NAME_MAX 32
 
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief A transport name and device name pair used by this endpoint.
+ */
+typedef struct ucp_transport_entry_t {
+    /**
+     * The name of a transport layer used by this endpoint.
+     */
+    const char *transport_name;
+
+    /**
+     * The name of the device used with this transport by this endpoint.
+     */
+    const char *device_name;
+} ucp_transport_entry_t;
+
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief Structure containing an array of transport layers and device names
+ * used by an endpoint.
+ *
+ * The caller is responsible for allocation and deallocation of 
+ * this structure.
+ */
+typedef struct ucp_transports {
+    /**
+     * Number of transport/device name pairs. The caller must set this to
+     * the maximum number of these pairs the structure can contain. On return,
+     * this is set to the actual number of transport/device pairs used by the
+     * endpoint.
+     */
+    unsigned              num_entries;
+
+    /**
+      * Pointer to array of transport/device name pairs used by this endpoint.
+      * The caller is responsible for the allocation and deallocation of this
+      * array.
+      */
+    ucp_transport_entry_t *entries;
```

Comment:
maybe caller should also provide the size of each "entry" struct, to allow extending it in the future with more fields.
@shamisp WDYT?

Reply:
@yosefe not sure if I understand what you propose here. `ucp_transport_entry_t` is defined by UCP, so we control the size. 

I'm concerned that both `ucp_transport_entry_t` and `ucp_transports_t` introduce unprotected fields which may lead to ABI backward compatibility issues in the future.

Reply:
since ucp_transport_entry_t is an array, user may want to pass the size of each entry in user array. if we add more fields over time, it will help maintain ABI compatibility (UCP will address each array element according to user-provided size)

Reply:
@yosefe Do you mean number of entries in `*entries` ? `ucp_transport_entry_t` is a struct with two fields (transport names and device names). Do you want user provides control over the size of the names ? I don't think this is enough. Maybe we will want to add some additional non-string members.

Reply:
@shamisp i mean something like https://github.com/linux-rdma/rdma-core/blob/ebbdb85d2dadbd8850e9647d7e3cfeb098569b35/libibverbs/verbs.h#L2379 (note 'entry_size')

Reply:
The intent of my changes are to return array of the transport name and device name pairs associated with a UCX endpoint. When my UCX code builds the array, it just copies the string pointers pointing to the transport name and device name that UCX set when the endpoint was set up. So there should not be any problems with the size of these strings from the UCX side. If the caller has size limits, then it is the caller's responsibility to use strncpy, etc.

If I understand, you are also asking that the uct_transport_t struct contain a bit flags field similar to that used in other UCX structures so the user specifies what fields in the struct he wants UCX to set or for UCX to look at.

The entire contents of the ucp_transports_t struct is already protected by the UCP_EP_FIELD_TRANSPORTS flag in the ucp_ep_attr flags. I''m intending for this struct to always return transport/device pairs. I suppose more fields could be added in the future if someone else wants them. But I'm wondering if individual flags to specify access to each field are needed.

Reply:
@yosefe so, is the intent that user will pass `sizeof(ucp_transport_entry_t)` ?

Reply:
@shamisp @yosefe I can make the other changes asked for once the question about adding bit fields to ucp_transport_t is resolved. I look at the ucp_transport_entry_t structure as a specific piece of data that the user requests and gets back in it's entirety if the user sets the UCP_EP_ATTR_FIELD_TRANSPORT bit in the ucp_transports_t structure's flags field, not a set of fields the user can selectively choose.

Reply:
@drwootton @yosefe - general note: IMHO this is useful API change. We just have to figure out the best way to express this in API.  I will not be able to join this week call, but I would suggest to discuss ideas on the upcoming call.

Reply:
@shamisp I can join the call  to discuss. If it's this week I will have to leave by 3PM NY time. Do i need to do anything to get this topic on the agenda or do i just connect via teams or call the phone number in the wiki and wait for the topic to be raised?

Reply:
I'll join the call to discuss, just connect to the teams link. agree with @shamisp this is a useful API, just need to figure out the backward compatibility method.

Reply:
@drwootton @yosefe you can propose the agenda on UCX mailing list https://elist.ornl.gov/mailman/listinfo/ucx-group
call info https://github.com/openucx/ucx/wiki/Bi-Weekly-conference-call